File: packages/sdk-go/offline_test.go
package ubag

import "testing"

func TestOfflineQueueRoundTrip(t *testing.T) {
	store := NewMemoryOfflineStore()
	q := NewOfflineQueue(store)
	q.Enqueue(map[string]any{"target": "mock"})
	q.Enqueue(map[string]any{"target": "mock"})
	if q.Size() != 2 {
		t.Fatalf("expected 2, got %d", q.Size())
	}
	var sent int
	err := q.Flush(func(req map[string]any) error { sent++; return nil })
	if err != nil {
		t.Fatalf("flush error: %v", err)
	}
	if sent != 2 || q.Size() != 0 {
		t.Fatalf("expected 2 sent and empty queue, got sent=%d size=%d", sent, q.Size())
	}
}

File: packages/sdk-typescript/test/offline.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import { OfflineQueue } from "../dist/offline.js";

function memoryAdapter() {
  let entries = [];
  return {
    read: () => entries,
    write: (e) => { entries = e; },
    _dump: () => entries,
  };
}

test("enqueue persists an entry", () => {
  const adapter = memoryAdapter();
  const q = new OfflineQueue(adapter);
  q.enqueue({ target: "mock", command_type: "submit", input: {} });
  assert.equal(adapter._dump().length, 1);
});

test("flush drains entries through the sender and clears the queue", async () => {
  const adapter = memoryAdapter();
  const q = new OfflineQueue(adapter);
  q.enqueue({ target: "mock", command_type: "submit", input: {} });
  q.enqueue({ target: "mock", command_type: "submit", input: {} });
  const sent = [];
  await q.flush(async (req) => { sent.push(req); });
  assert.equal(sent.length, 2);
  assert.equal(adapter._dump().length, 0);
});

test("flush stops and retains entries when the sender throws", async () => {
  const adapter = memoryAdapter();
  const q = new OfflineQueue(adapter);
  q.enqueue({ target: "mock", command_type: "submit", input: {} });
  await assert.rejects(q.flush(async () => { throw new Error("offline"); }));
  assert.equal(adapter._dump().length, 1);
});
